crates/burn-core/src/module/display.rs

```diff
@@ -0,0 +1,543 @@
+        let settings = if let Some(custom_settings) = self.custom_settings() {
+            custom_settings.inherit(passed_settings)
+        } else {
+            passed_settings
+        };
+
+        let indent = " ".repeat(settings.level * settings.indentation_size());
+        let indent_close_braces = " ".repeat((settings.level - 1) * settings.indentation_size());
+
+        let settings = settings.level_up();
+
+        let self_type = extract_type_name::<Self>();
+
+        // Use custom content if it is implemented and show_all_attributes is false,
+        // otherwise use default content
+        let content = if !settings.show_all_attributes() {
+            self.custom_content(Content::new(settings.clone()))
+                .unwrap_or_else(|| {
+                    self.content(Content::new(settings.clone()))
+                        .unwrap_or_else(|| {
+                            panic!("Default content should be implemented for {self_type}.")
+                        })
+                })
+        } else {
+            self.content(Content::new(settings.clone()))
+                .unwrap_or_else(|| panic!("Default content should be implemented for {self_type}."))
+        };
+
+        let top_level_type = if let Some(top_level_type) = content.top_level_type {
+            top_level_type.to_owned()
+        } else {
+            self_type.to_owned()
+        };
+
+        // If there is only one item in the content, return it or no attributes
+        if let Some(item) = content.single_item {
+            return item;
+        } else if content.attributes.is_empty() {
+            return top_level_type.to_string();
+        }
+
+        let mut result = String::new();
+
+        // Print the struct name
+        if settings.new_line_after_attribute() {
+            writeln!(result, "{} {{", top_level_type).unwrap();
+        } else {
+            write!(result, "{} {{", top_level_type).unwrap();
+        }
+
+        for (i, attribute) in content.attributes.iter().enumerate() {
+            if settings.new_line_after_attribute() {
+                writeln!(result, "{indent}{}: {}", attribute.name, attribute.value).unwrap();
+            } else if i == 0 {
+                write!(result, "{}: {}", attribute.name, attribute.value).unwrap();
+            } else {
+                write!(result, ", {}: {}", attribute.name, attribute.value).unwrap();
+            }
+        }
+
+        if settings.show_num_parameters() {
+            if settings.new_line_after_attribute() {
+                writeln!(result, "{indent}params: {}", self.num_params()).unwrap();
+            } else {
+                write!(result, ", params: {}", self.num_params()).unwrap();
+            }
+        }
+        if settings.new_line_after_attribute() {
+            write!(result, "{indent_close_braces}}}").unwrap();
+        } else {
+            write!(result, "}}").unwrap();
+        }
+
+        result
```

Comment:
I'm unsure, but maybe we should extract this into a function that we call inside the default implementation. This way, the function could be compiled only once and reused for all modules. In the current case, I think the compiler is generating that function for all types, creating a bigger binary with slower compile time. This is just an intuition and it is definitely hard to measure, but I think it's not a bad thing to extract this into one function.

Reply:
To clarify and confirm I asked these question on Rust forum:

In Rust, when implementing a default method as part of a trait that can be implemented multiple times, is it true that:

1. The compiler generates this function separately for all types implementing the trait?
2. This approach potentially leads to a larger binary size and slower compile times?
3. Extracting the common functionality into a separate function and calling it from the default trait implementation could optimize this?
4. Such an extraction would allow the function to be compiled only once and reused across all modules?

If these points are accurate, what are the trade-offs to consider when deciding whether to implement this optimization?

Reply (https://discord.com/channels/442252698964721669/443150878111694848/1253892142267498599):

1. is true if you dont provide another implementation of the function
2. is not really true, the compiler merges identical functions so it wont increase the binary size unless the function does different things depending on the type
3. wouldnt really help because of 2, unless you extract only the identical stuff into a function and left the different stuff in the default impl
4. yes, if the final code is identical for all types

So in the end, I do not think it makes sense to extract the method because the function does behave differently depending on the Self type. Even if I would extract, it would be pretty much the same: self needs to be passed.